gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -219,6 +486,59 @@ var fluid = fluid || require("infusion");
         return promiseTogo;
     };
 
+    /**
+     * Stand alone method to add a `gpiiKey` document in the system. `prefsSafeId` and `prefSetId` are
+     * required fields. Optionally you can specify the `gpiiKey` otherwise a new one will be generated.
+     * This key must not exist yet.
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param {Object} keyData - Data for the new `gpiiKey` document.
+     * @return {Promise} A promise object that resolves to the new saved document, or rejects with a
+     * standard error payload.
+     */
+    gpii.preferencesServer.preferencesService.addGpiiKey = function (that, dataStore, keyData) {
+        var promTogo = fluid.promise();
+
+        if (!keyData.prefsSafeId || !keyData.prefsSetId) {
+            promTogo.reject({
+                isError: true,
+                message: "Unable to add GPII Key to Preferences Safe. `prefsSafeId` and `prefsSetId` are required fields."
+            });
+        }
+        else {
+            // Preferences Safe must exist to add this key
+            that.getPrefsSafe(keyData.prefsSafeId).then(function () {
```

Comment:
This else block seems a good use case of `fluid.promise.fireTransformEvent()`?

@amb26, I'm also confused about when `fluid.promise.fireTransformEvent()` should be used and when a chain of then() should be used. Your advice is appreciated.

Reply:
It's interesting.  I think it makes sense for grouping together distinct operations... but it can be a little cumbersome when you need to add a precondition such as just checking for the existence of a prefssafe before continuing forward.

Maybe improving the tooling and streamlining the syntax for these chained operations and preconditions could be a BOF in Toronto next month.

Reply:
I agree that this in-function promise chain is too long and should be broken up